src/curve.rs

```diff
@@ -51,89 +30,154 @@ impl<F: Field, C: EdwardsCurve<F>> Clone for EdwardsPoint<F, C> {
-        builder.build().execute(&mut values);
-        new_point.evaluate(&values)
-    }
-
-    /// Given an `x` and `y` coordinate, checks that they constitute a point on the curve
-    /// and returns an `EdwardsPoint`
-    pub fn from_elements(x: Element<F>, y: Element<F>) -> EdwardsPoint<F, C> {
+    pub fn new(x: Element<F>, y: Element<F>) -> EdwardsPoint<F, C> {
         assert!(C::a() * &x * &x + &y * &y == Element::one() + C::d() * &x * &x * &y * &y,
                 "Point must be contained on the curve.");
         EdwardsPoint { x, y, phantom: PhantomData }
     }
 
-    /// Returns the Y coordinate of an `EdwardsPoint`
-    pub fn compressed(&self) -> &Element<F> {
+    pub fn compressed_element(&self) -> &Element<F> {
         &self.y
     }
 }
 
+pub struct EdwardsExpression<F: Field, C: EdwardsCurve<F>> {
+    pub x: Expression<F>,
+    pub y: Expression<F>,
+    phantom: PhantomData<*const C>,
+}
 
-impl<F: Field, C: EdwardsCurve<F>> EdwardsPointExpression<F, C> {
-    /// Returns the Y coordinate of an `EdwardsPointExpression`
-    pub fn compressed(&self) -> &Expression<F> {
-        &self.y
+impl<F: Field, C: EdwardsCurve<F>> EdwardsExpression<F, C> {
+    pub fn new(
+        builder: &mut GadgetBuilder<F>,
+        x: Expression<F>,
+        y: Expression<F>,
+    ) -> EdwardsExpression<F, C> {
+        let x_squared = builder.product(&x, &x);
+        let y_squared = builder.product(&y, &y);
+        let x_squared_y_squared = builder.product(&x_squared, &y_squared);
+        builder.assert_equal(&(&x_squared * C::a() + &y_squared),
+                             &(&x_squared_y_squared * C::d() + Expression::one()));
+        EdwardsExpression::new_unsafe(x, y)
+    }
+
+    pub fn new_unsafe(x: Expression<F>, y: Expression<F>) -> EdwardsExpression<F, C> {
+        EdwardsExpression { x, y, phantom: PhantomData }
+    }
+}
+
+impl<F: Field, C: EdwardsCurve<F>> GroupExpression<F> for EdwardsExpression<F, C> {
+    fn compressed_expression(&self) -> &Expression<F> { &self.y }
+    fn to_component_expression(&self) -> Vec<Expression<F>> { vec![self.x.clone(), self.y.clone()] }
+
+    fn from_component_expression_unsafe(components: Vec<Expression<F>>) -> Self {
+        Self::new_unsafe(components[0].clone(), components[1].clone())
+    }
+}
+
+impl<F: Field, C: EdwardsCurve<F>> From<&EdwardsPoint<F, C>> for EdwardsExpression<F, C> {
+    fn from(point: &EdwardsPoint<F, C>) -> Self {
+        EdwardsExpression {
+            x: Expression::from(&point.x),
+            y: Expression::from(&point.y),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<F: Field, C: EdwardsCurve<F>> From<(Element<F>, Element<F>)> for EdwardsExpression<F, C> {
+    fn from(coordinates: (Element<F>, Element<F>)) -> Self {
+        let point = EdwardsPoint::new(coordinates.0, coordinates.1);
+        EdwardsExpression::from(&point)
+    }
+}
+
+impl<F: Field, C: EdwardsCurve<F>> From<Vec<&Expression<F>>> for EdwardsExpression<F, C> {
```

Comment:
Can we remove this now? Or if we keep it, I think it's best to remove the reference in the Vec